core/src/mast/node/basic_block_node/op_batch.rs

```diff
@@ -11,20 +11,55 @@ use super::{BATCH_SIZE, Felt, GROUP_SIZE, Operation, ZERO};
 /// operations or a single immediate value.
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct OpBatch {
+    /// A list of operations in this batch, including padding noops.
     pub(super) ops: Vec<Operation>,
+    /// An array of indexes in the ops array, marking the beginning and end of each group.
+    ///
+    /// The array maintains the invariant that the i-th group (i <= BATCH_SIZE-1) is at
+    /// `self.ops[self.indptr[i]..self.indptr[i+1]]`.
+    ///
+    /// By convention, the groups containing immediate values have a zero-length slice of the ops
+    /// array.
+    pub(super) indptr: [usize; Self::BATCH_SIZE_PLUS_ONE],
+    /// An array of bits representing whether a group had undergone padding
+    pub(super) padding: [bool; BATCH_SIZE],
+    /// Value of groups in the batch, which includes operations and immediate values.
     pub(super) groups: [Felt; BATCH_SIZE],
-    pub(super) op_counts: [usize; BATCH_SIZE],
+    /// Number of groups in this batch.
+    ///
+    /// The arrays above are meaningful in their [0..self.num_groups] prefix
+    /// (or [0..self.num_groups + 1] in the case of the indptr array).
     pub(super) num_groups: usize,
```

Comment:
At some point in the future (after we get rid of the slow processor), I wonder if we could get rid of these fields and maybe keep only the `ops` field as I think this is really the only field we need to execute programs in the fast processor.

Or, if we wanted to optimize this even further, we could have a single vector of operations for the entire basic block. It could look something like this:

```rust
pub struct BasicBlockNode {
    /// operations here would contain Respan operations to mark boundaries of operation batches
    ops: Vec<Operation>,
    digest: Word,
}
```

This structure would make serialization/deserialization and hash computations more difficult, but would make executing programs via the fast processor as fast as possible.

Reply:
I definitely think that would be a lot easier to reason about, and the perf boost around raw execution would probably be significant. I don't think serialization/deserialization is much of a concern (i.e. it is fine if they are complex, since it is relatively rare), but since hashing is important for performance, it isn't clear to me whether it would be a net win or not. I think as long as it breaks even, or is close enough to not matter, then it would be nicer this way.